### PR TITLE
ref: (Django 1.9) Bump djangorestframework to 3.0.5 as an intermediate step to get to 3.3.x

### PR DESCRIPTION
**Pull request by wedamija**
Upgrade to next major version of django rest framework. This diff is fairly large, but this is the minimum set of changes I could make to have the upgrade pass. Mostly things are drop in renames, so there aren't really sweeping changes here.

Summary of changes here: https://www.django-rest-framework.org/community/3.0-announcement/

Getsentry pr here: https://github.com/getsentry/getsentry/pull/3009
sentry-plugins doesn't need changes from what I can tell.

Things done:
- `serializer.object` replaced with `serializer.validated_data`. For standard serializers this is a drop in replacement. For model serializers we get a dictionary rather than the instance, so need to translate usage accordingly.
- on serializer fields, replace `allow_none` with `allow_null`
- on field validation methods in serializers (ie `validate_<fieldname>`) change parameters to just `value`, rather than `attrs, source`. This means that values are just passed directly to the validation method rather than us having to look them up. In cases where we use `attrs` to look up other values we do so in `validate`
- replaced uses of `init_data` with `initial_data` on serializers
- add `allow_blank` and `allow_null` as necessary to replicate previous behaviour. These need to be specified explicitly now.
- Replace uses of `Serializer.base_fields` with `Serializer().fields`. These are equivalent
- Convert all `WritableField` classes to `Field`. Fields are now writeable by default.
- Replace our vendorized `ListField` with the DRF version
- For custom fields, rename `from_native` to `to_internal_value`, and `to_native` to `to_representation`. These are drop in replacement
- `get_default_fields` has been removed. `get_fields` works as a replacement, but doesn't operate in quite the same way, so restructed the validators for monitors.

### CODE REVIEW

File: src/sentry/api/endpoints/project_user_reports.py [hunk 1]
```diff
@@ -185,38 +186,38 @@ def post(self, request, project):
...
             )
         except Event.DoesNotExist:
-            if not report.email:
+            if not report_data.get('email'):
                 return None
             try:
                 return EventUser.objects.filter(
-                    project_id=report.project_id,
-                    email=report.email,
+                    project_id=report_data['project'].id,
```
Comment (markstory) [inline src/sentry/api/endpoints/project_user_reports.py]: Does the report no longer have a `project_id` attribute?

File: src/sentry/api/validators/monitor.py [hunk 2]
```diff
@@ -69,7 +74,7 @@ def validate_schedule(self, attrs, source):
             if not isinstance(value[0], int):
                 raise ValidationError('Invalid value for schedule unit count (index 0)')
             if value[1] not in INTERVAL_NAMES:
-                raise ValidationError('Invalid value for schedlue unit name (index 1)')
+                raise ValidationError('Invalid value for schedule unit name (index 1)')
```
Comment (markstory) [inline src/sentry/api/validators/monitor.py]: :clap:

File: src/sentry/api/validators/monitor.py [hunk 3]
```diff
@@ -96,27 +101,46 @@ class MonitorValidator(serializers.Serializer):
...
+    def validate_type(self, value):
         if value:
-            attrs[source] = MONITOR_TYPES[value]
-        return attrs
+            value = MONITOR_TYPES[value]
+        return value
+
+    def update(self, instance, validated_data):
+        has_config = 'config' in instance or 'config' in validated_data
+        if 'config' in instance or 'config' in validated_data:
```
Comment (markstory) [inline src/sentry/api/validators/monitor.py]:
```suggestion
        if has_config:
```

Seems like the same logic here twice.

File: tests/sentry/api/endpoints/test_sudo.py [hunk 4]
```diff
@@ -29,9 +29,9 @@ def test_sudo_required_del_org(self):
         with self.settings(MIDDLEWARE_CLASSES=tuple(middleware)):
             response = self.client.delete(url, is_sudo=False)
             assert response.status_code == 401
-            assert response.data['detail']['code'] == 'sudo-required'
-            assert response.data['detail']['message'] == 'Account verification required.'
-            assert response.data['detail']['extra']['username'] == 'foo@example.com'
+            assert response.data['code'] == 'sudo-required'
+            assert response.data['message'] == 'Account verification required.'
+            assert response.data['extra']['username'] == 'foo@example.com'
```
Comment (markstory) [inline tests/sentry/api/endpoints/test_sudo.py]: Does the frontend need to be updated for these changes?

File: src/sentry/api/endpoints/project_user_reports.py (hunk 1 above)
Comment (wedamija) [reply src/sentry/api/endpoints/project_user_reports.py]: It still does, we just no long have a `UserReport` model instance at this point, we just have the dictionary from `serializer.validated_data`

File: tests/sentry/api/endpoints/test_sudo.py (hunk 4 above)
Comment (wedamija) [reply tests/sentry/api/endpoints/test_sudo.py]: Damn, yeah. I'll see if I can figure out why this change is actually happening. For most api responses the `detail` level is maintained, but not for this middleware.

Comment (wedamija) [reply tests/sentry/api/endpoints/test_sudo.py]: @markstory Resolved this, good catch. My changes should keep our responses the same as they were before.

**Review by markstory — APPROVED** [review]

**Pull request merged**